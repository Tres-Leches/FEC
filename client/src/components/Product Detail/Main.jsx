import React from 'react';
import axios from 'axios';
import './productDetail.css';

import Add2Cart from './Add2Cart';
import DefaultView from './DefaultView';
import ExpandedView from './ExpandedView';
import ProductInfo from './ProductInfo';
import StyleSelector from './StyleSelector';
import Share from './Share';
import EndInfo from './EndInfo';

class Main extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      product_ids: [16392, 16465, 16056, 16084, 16154, 16060, 16072, 16073],
      product_id: 16060,
      zoomed: false,
      product: null,
      styles: [],
      style: null,
      mainPhoto: null,
    };
  }

  componentDidMount() {
    axios.get(`/api/products/${this.state.product_id}/styles`)
      .then((response) => {
        this.setState({styles: response.data.results, style: response.data.results[0], mainPhoto: response.data.results[0].photos[0]})
      })
      .catch(err => console.error(err))
    axios.get(`/api/products/${this.state.product_id}`)
      .then((response) => {
        this.setState({product: response.data})
      })
      .catch(err => console.error(err))
  }

  changeStyle(style) {
    this.setState({style})
  }

  changeMainPhoto (photo) {
    this.setState({mainPhoto: photo})
  }

  changeView() {
    this.setState({zoomed: !this.state.zoomed})

  }

  render() {
    if(this.state.style && this.state.product){
      let expand;
      if (this.state.zoomed) {
        expand = <ExpandedView style={this.state.style}
          mainPhoto={this.state.mainPhoto}
          changeMainPhoto={this.changeMainPhoto.bind(this)}
          changeView={this.changeView.bind(this)}/> ;
        document.body.style.overflow = 'hidden';
      } else {
        expand = "";
        document.body.style.overflow = 'visible';
      }
      return (
        <div>
          {expand}
          <div className="overview">

            <div className="view">
              <DefaultView style={this.state.style} changeView={this.changeView.bind(this)} mainPhoto={this.state.mainPhoto} changeMainPhoto={this.changeMainPhoto.bind(this)}/>
              <div className="info">
              <ProductInfo product={this.state.product} productId={this.state.product_id}/>
              <StyleSelector styles={this.state.styles} style={this.state.style} changeStyle={this.changeStyle.bind(this)}/>
              <Add2Cart style={this.state.style}/>
              <Share />
            </div>
            </div>

            <div className="viewLastInfo">
              <EndInfo product={this.state.product}/>
            </div>
          </div>
        </div>
      );
    } else {
      return (
        <div></div>
      )
    }
  }
}

export default Main;
