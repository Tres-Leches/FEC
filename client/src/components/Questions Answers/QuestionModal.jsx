/* eslint-disable react/prop-types */
import React from 'react';
import axios from 'axios';
import * as yup from 'yup';

const schema = yup.object().shape({
  body: yup.string().required().max(1000),
  name: yup.string().required().max(60),
  email: yup.string().required().max(60).email(),
  product_id: yup.number().positive().integer(),
});

class QuestionModal extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      userQuestion: '',
      userNickname: '',
      userEmail: '',
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {
    const { value } = e.target;
    const { name } = e.target;
    this.setState({
      [name]: value,
    });
  }

  handleSubmit(e) {
    const { closeModal, productId, getQuestions } = this.props;
    const { userQuestion, userNickname, userEmail } = this.state;
    e.preventDefault();
    schema
      .validate({
        body: userQuestion,
        name: userNickname,
        email: userEmail,
        product_id: Number(productId),
      })
      .then((value) => {
        axios.post(`/api/qa/questions/${productId}`, value)
          .then(() => {
            getQuestions();
          })
          .catch((err) => console.error(err));
      })
      .then(() => closeModal())
      .catch((err) => (window.alert(err)));
  }

  render() {
    const { show, closeModal } = this.props;
    const { userQuestion, userNickname, userEmail } = this.state;

    if (!show) {
      return null;
    }
    return (
      <div className="qa-modal">
        <div className="qa-modal-content">
          <div className="qa-modal-header">
            <h2>Post your question</h2>
            <span
              className="close"
              role="button"
              tabIndex={0}
              onClick={closeModal}
              onKeyDown={closeModal}
            >
              &times;
            </span>
          </div>
          <form onSubmit={this.handleSubmit}>
            <label htmlFor="user-question">
              <span className="label">Your Question </span>
              <span className="required">*</span>
              <textarea
                name="userQuestion"
                placeholder=" What is your question?"
                value={userQuestion}
                onChange={this.handleChange}
              />
            </label>
            <label htmlFor="user-nickname">
              <span className="label">Nickname </span>
              <span className="required">*</span>
              <input
                type="text"
                name="userNickname"
                placeholder=" Example: jackson11"
                value={userNickname}
                onChange={this.handleChange}
              />
              <span className="disclaimer">For privacy reasons, do not use your full name or email address.</span>
            </label>
            <label htmlFor="user-email">
              <span className="label">Email </span>
              <span className="required">*</span>
              <input
                type="text"
                name="userEmail"
                placeholder=" Example: jackson@hackreactor.com"
                value={userEmail}
                onChange={this.handleChange}
              />
              <span className="disclaimer">For privacy reasons, do not use your full name or email address.</span>
            </label>
            <label htmlFor="form-actions">
              <input type="button" onClick={closeModal} value="Cancel" />
              <input type="submit" value="Post" />
            </label>
          </form>
          <div className="qa-modal-footer" />
        </div>
      </div>
    );
  }
}

export default QuestionModal;
