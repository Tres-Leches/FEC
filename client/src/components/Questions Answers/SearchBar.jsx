/* eslint-disable react/prop-types */

import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSearch } from '@fortawesome/free-solid-svg-icons';

const element = <FontAwesomeIcon icon={faSearch} />;

const SearchBar = ({ searchQuery, setSearchQuery, resetSearchQuery }) => (
  <div className="searchbar-wrapper">
    <form className="searchbar-form">
      <div id="icon">{element}</div>
      <input id="search-field" type="text" value={searchQuery || ''} placeholder="HAVE A QUESTION? SEARCH FOR ANSWERS..." onChange={setSearchQuery} />
      <span
        className="search-reset"
        role="button"
        tabIndex={0}
        onClick={resetSearchQuery}
        onKeyDown={resetSearchQuery}
      >
        &times;
      </span>
    </form>
  </div>
);

export default SearchBar;
