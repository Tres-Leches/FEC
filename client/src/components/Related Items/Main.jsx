import React from 'react';

class Main extends React.Component {
  constructor(props) {
    super(props);
    this.state = {

    };
  }

  render() {
    return (
      <div>
        Hello from Related Items
      </div>
    );
  }
}

export default Main;
