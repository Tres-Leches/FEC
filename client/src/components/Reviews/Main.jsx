/* eslint-disable no-console */
import React from 'react';
import axios from 'axios';
import ReviewList from './ReviewList.jsx';
import MetaData from './MetaData.jsx';
import './reviews.css';

class Main extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      productId: this.props.productId,
      product: this.props.product,
      reviews: '',
      metaData: '',
    };
  }

  componentDidMount() {
    Promise.all([
      axios.get(`/api/reviews/${this.state.productId}/relevant`),
      axios.get(`/api/reviews/meta/${this.state.productId}`),
    ]).then(([res1, res2]) => {
      this.setState({
        reviews: res1.data,
        metaData: res2.data,
      }, () => console.log(this.state));
    });
  }

  getReviews(sort) {
    axios.get(`/api/reviews/${this.state.productId}/${sort}`)
      .then((res) => {
        this.setState({
          reviews: res.data,
        });
      }).then(() => console.log(this.state)).catch((err) => (console.log(err)));
  }

  render() {
    if (!this.state.reviews || !this.state.metaData) {
      return (
        <div />
      );
    }
    return (
      <div>
        <div>Ratings &amp; Reviews</div>
        <MetaData metaData={this.state.metaData} />
        <ReviewList
          reviews={this.state.reviews}
          meta={this.state.metaData}
          getReviews={this.getReviews.bind(this)}
        />
      </div>
    );
  }
}

export default Main;
