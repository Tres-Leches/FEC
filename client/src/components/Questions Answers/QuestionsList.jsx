/* eslint-disable react/prop-types */
import React from 'react';
import Question from './Question';
import QuestionModal from './QuestionModal';

class QuestionsList extends React.Component {
  constructor(props) {
    super(props);
    const { questions } = this.props;
    this.state = {
      show: false,
      questionsLen: 2,
      remainLen: questions.length - 2,
      hasMoreItems: true,
    };
    this.showModal = this.showModal.bind(this);
    this.closeModal = this.closeModal.bind(this);
    this.updateQuestionLen = this.updateQuestionLen.bind(this);
    this.resetQuestionLen = this.resetQuestionLen.bind(this);
    this.handleListScroll = this.handleListScroll.bind(this);
  }

  componentDidUpdate(prevProps) {
    const { questions } = this.props;
    if (questions !== prevProps.questions) {
      this.resetQuestionLen();
    }
  }

  handleListScroll(e) {
    const { hasMoreItems } = this.state;
    const bottom = e.target.scrollHeight - e.target.scrollTop === e.target.clientHeight;
    if (bottom) {
      if (hasMoreItems) {
        this.updateQuestionLen();
      }
    }
  }

  updateQuestionLen() {
    const { questionsLen, remainLen } = this.state;
    this.setState({
      questionsLen: questionsLen + 2,
      remainLen: remainLen - 2,
    }, () => {
      if (remainLen <= 0) {
        this.setState({ hasMoreItems: false });
      }
    });
  }

  resetQuestionLen() {
    const { questions } = this.props;
    this.setState({
      questionsLen: 2,
      remainLen: questions.length - 2,
      hasMoreItems: true,
    });
  }

  showModal() {
    this.setState({ show: true }, () => {
      const { show } = this.state;
      if (show) {
        document.body.style.overflow = 'hidden';
        document.getElementsByClassName('scroll').overflow = 'hidden';
      }
    });
  }

  closeModal() {
    this.setState({ show: false }, () => {
      const { show } = this.state;
      if (!show) {
        document.body.style.overflow = 'unset';
        document.getElementsByClassName('scroll').overflow = 'unset';
      }
    });
  }

  render() {
    const { questions, getQuestions, productId } = this.props;
    const { show, questionsLen, remainLen } = this.state;

    if (questions) {
      return (
        <div className="questionslist-container">
          <div className="scroll" onScroll={this.handleListScroll}>
            {questions.slice(0, questionsLen).map((question) => (
              <Question
                key={question.question_id}
                question={question}
                getQuestions={getQuestions}
              />
            ))}
          </div>
          {(remainLen > 0) && (
            <button
              type="button"
              className="questionslist-actions"
              onClick={this.updateQuestionLen}
            >
              {`MORE ANSWERED QUESTIONS (${remainLen})`}
            </button>
          )}
          {(remainLen <= 0 && questions.length > 2) && (
            <button
              type="button"
              className="questionslist-actions"
              onClick={this.resetQuestionLen}
            >
              COLLAPSE QUESTIONS
            </button>
          )}
          <button
            type="button"
            className="questionslist-actions"
            onClick={this.showModal}
          >
            ADD A QUESTION +
          </button>
          <QuestionModal
            show={show}
            showModal={this.showModal}
            closeModal={this.closeModal}
            productId={productId}
            getQuestions={getQuestions}
          />
        </div>
      );
    }
    return null;
  }
}

export default QuestionsList;
