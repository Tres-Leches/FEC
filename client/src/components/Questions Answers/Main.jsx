/* eslint-disable no-console */
import React from 'react';
import axios from 'axios';
import _ from 'underscore';
import SearchBar from './SearchBar';
import QuestionsList from './QuestionsList';

class Main extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      questions: [],
      filteredQuestions: [],
      searchQuery: '',
    };
    this.getQuestions = this.getQuestions.bind(this);
    this.setSearchQuery = this.setSearchQuery.bind(this);
    this.filterQuestions = this.filterQuestions.bind(this);
  }

  componentDidMount() {
    this.getQuestions();
  }

  setSearchQuery(e) {
    this.setState({ searchQuery: e.target.value }, () => {
      this.filterQuestions();
    });
  }

  getQuestions() {
    const productId = 16392; // Placeholder
    axios.get(`/api/qa/questions/${productId}`)
      .then((response) => {
        const questions = response.data.results;
        questions.sort((a, b) => (
          b.question_helpfulness - a.question_helpfulness
        ));
        this.setState({ questions });
      })
      .then(() => { this.filterQuestions(); })
      .catch((err) => console.error(err));
  }

  filterQuestions() {
    const { searchQuery, questions } = this.state;
    let filtered = [];
    if (searchQuery.length >= 3) {
      questions.forEach((question) => {
        const query = searchQuery.toLowerCase();
        const qBody = question.question_body.toLowerCase();
        if (qBody.includes(query)) {
          filtered.push(question);
        } else {
          // Else search through answers of each question_id.
          // If answer.body, aBody, includes query, return question
          const { answers } = question;
          Object.keys(answers).forEach((answerId) => {
            const aBody = question.answers[`${answerId}`].body.toLowerCase();
            if (aBody.includes(query)) {
              if (_.findWhere(filtered, aBody) == null) {
                filtered.push(question);
              }
            }
          });
        }
      });
    } else {
      filtered = questions;
    }
    this.setState({ filteredQuestions: filtered });
  }

  render() {
    const { filteredQuestions, searchQuery } = this.state;

    return (
      <div>
        Hello from Questions and Answers
        <SearchBar
          searchQuery={searchQuery}
          setSearchQuery={this.setSearchQuery}
          handleSearchSubmit={this.handleSearchSubmit}
        />
        <QuestionsList questions={filteredQuestions} />
      </div>
    );
  }
}

export default Main;
