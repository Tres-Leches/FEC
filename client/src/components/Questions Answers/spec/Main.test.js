/* eslint-disable no-undef */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { shallow, mount, render } from 'enzyme';
import Main from '../Main';
import SearchBar from '../SearchBar';
import QuestionsList from '../QuestionsList';
import QuestionModal from '../QuestionModal';

const questions = [{
  question_id: 98125,
  question_body: 'Odio voluptatum qui autem fuga ullam doloremque similique aut architecto.',
  question_date: '2020-11-24T00:00:00.000Z',
  asker_name: 'Misael32',
  question_helpfulness: 48,
  reported: false,
  answers: {
    930659: {
      id: 930659,
      body: 'Alias veniam ut.',
      date: '2020-08-02T00:00:00.000Z',
      answerer_name: 'Sammie.Hudson',
      helpfulness: 16,
      photos: [
        'https://images.unsplash.com/photo-1542272454315-4c01d7abdf4a?ixlib=rb-1.2.1&auto=format&fit=crop&w=2850&q=80',
      ],
    },
  },
}];

describe('Main', () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<Main debug />);
    expect(component).toMatchSnapshot();
  });
});

describe('Render Search Bar', () => {
  it('should render search bar', () => {
    const component = shallow(<SearchBar />);
    expect(component).toMatchSnapshot();
  });

  it('should take search parameter as prop', () => {
    const query = 'string';
    const component = mount(<SearchBar searchQuery={query} />);
    expect(component.props().searchQuery).toEqual(query);
  });

  it('should change state after input change', () => {
    const query = 'string';
    const wrapper = mount(<Main />);
    const input = wrapper.find('#search-field');
    expect(input).toHaveLength(1);
    input.simulate('change', { target: { value: query } });
    expect(wrapper.state().searchQuery).toEqual(query);
  });
});

describe('Render Questions List', () => {
  it('should render questions list', () => {
    const component = shallow(<QuestionsList questions={questions} />);
    expect(component).toMatchSnapshot();
  });

  it('should take questions array as prop', () => {
    const component = mount(<QuestionsList questions={questions} />);
    expect(component.props().questions).toEqual(questions);
  });

  it('should change state when toggled', () => {
    const wrapper = mount(<QuestionsList questions={questions} />);
    const button = wrapper.findWhere((node) => (
      node.type() === 'button' &&
      node.text() === 'ADD A QUESTION +'
    ));
    expect(button).toHaveLength(1);
    button.simulate('click');
    expect(wrapper.state().show).toEqual(true);
  });
});

describe('Render Question Modal', () => {
  it('should render question modal', () => {
    const component = shallow(<QuestionModal show />);
    expect(component).toMatchSnapshot();
  });

  it('should close question modal', () => {
    let show = true;
    const closeModal = () => {
      show = !show;
    };
    const wrapper = mount(<QuestionModal show closeModal={closeModal} />);
    const button = wrapper.find('.close');
    expect(button).toHaveLength(1);
    button.simulate('click');
    expect(show).toEqual(false);
  });

  it('should expect the correct inputs and submit button', () => {
    const show = true;
    const wrapper = mount(<QuestionModal show={show} />);
    const inputs = wrapper.find('form').find('input');
    expect(inputs).toHaveLength(4);
  });
});
