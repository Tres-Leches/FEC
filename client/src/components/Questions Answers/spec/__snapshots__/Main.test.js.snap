// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;

exports[`Render Question Modal should render question modal 1`] = `ShallowWrapper {}`;

exports[`Render Questions List should render questions list 1`] = `ShallowWrapper {}`;

exports[`Render Search Bar should render search bar 1`] = `ShallowWrapper {}`;
